apps/studio/src/server/modules/site/site.router.ts

```diff
@@ -328,4 +339,66 @@ export const siteRouter = router({
 
       return createSite({ siteName })
     }),
+  publishOne: protectedProcedure
+    .input(publishOneSiteSchema)
+    .mutation(async ({ ctx, input: { siteId } }) => {
+      await validateUserIsIsomerAdmin({ userId: ctx.user.id, gb: ctx.gb })
+
+      const byUser = await db
+        .selectFrom("User")
+        .selectAll()
+        .where("id", "=", ctx.user.id)
+        .executeTakeFirstOrThrow(
+          () =>
+            new TRPCError({
+              code: "NOT_FOUND",
+              message: "The user could not be found.",
+            }),
+        )
+
+      return db.transaction().execute(async (tx) => {
+        await logPublishEvent(tx, {
+          by: byUser,
+          eventType: AuditLogEvent.Publish,
+          delta: { before: null, after: null },
+          metadata: {},
+        })
+
+        await publishSite(ctx.logger, siteId)
+      })
+    }),
+  publishAll: protectedProcedure.mutation(async ({ ctx }) => {
+    await validateUserIsIsomerAdmin({ userId: ctx.user.id, gb: ctx.gb })
+
+    const byUser = await db
+      .selectFrom("User")
+      .selectAll()
+      .where("id", "=", ctx.user.id)
+      .executeTakeFirstOrThrow(
+        () =>
+          new TRPCError({
+            code: "NOT_FOUND",
+            message: "The user could not be found.",
+          }),
+      )
+
+    const sites = await db
+      .selectFrom("Site")
+      .selectAll()
+      .where("codeBuildId", "is not", null)
+      .execute()
+
+    return db.transaction().execute(async (tx) => {
```

Comment:
[nit] i'm unsure of how important this is, given that the # of sites we have is low and unlikely to exceed 1000 

but it's important to note that this might hold the transaction for an extended period of time and be potentially harmful to transactions that need to update audit log/version. i think a quick and easy fix would just be to use a `map` since order doesn't really matter here and we can do a `Promise.all` to avoid holding an extended lock.

wdyt?

Reply:
yeah i think it makes sense to not put in transaction since in this use case, it doesn't matter even if they it throw an error after triggering build for only a portion of the sites - will fix it!

Reply:
updated in https://github.com/opengovsg/isomer/pull/1242/commits/5508ae50874b0bded944d1879a48565a9c50ce9a